src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,316 @@
+         *
+         * @return KNNEngine associated with index allocation
+         */
+        public KNNEngine getKnnEngine() {
+            return knnEngine;
+        }
+
+        /**
+         * Getter for the path to the file from which the index was loaded.
+         *
+         * @return indexPath to index
+         */
+        public String getIndexPath() {
+            return indexPath;
+        }
+
+        /**
+         * Getter for the OpenSearch index associated with the native index.
+         *
+         * @return OpenSearch index name
+         */
+        public String getOsIndexName() {
+            return osIndexName;
+        }
+    }
+
+    /**
+     * Represents training data that has been allocated in native memory.
+     */
+    class TrainingDataAllocation implements NativeMemoryAllocation {
+
+        private volatile boolean closed;
+        private long pointer;
+        private final long size;
+
+        // Implement reader/writer with semaphores to deal with passing lock conditions between threads
+        private int readCount;
+        private Semaphore readSemaphore;
+        private Semaphore wrtieSemaphore;
+
+        /**
+         * Constructor
+         *
+         * @param pointer pointer in memory to the training data allocation
+         * @param size amount memory needed for allocation in kilobytes
+         */
+        public TrainingDataAllocation(long pointer, long size) {
+            this.closed = false;
+            this.pointer = pointer;
+            this.size = size;
+
+            this.readCount = 0;
+            this.readSemaphore = new Semaphore(1);
+            this.wrtieSemaphore = new Semaphore(1);
+        }
+
+        @Override
+        public void close() {
+            writeLock();
+
+            try {
+                if (closed) {
+                    return;
+                }
+
+                JNIService.freeVectors(this.pointer);
+            } finally {
+                closed = true;
+                writeUnlock();
+            }
+        }
+
+        @Override
+        public boolean isClosed() {
+            return closed;
+        }
+
+        @Override
+        public long getPointer() {
+            return pointer;
+        }
+
+        @Override
+        public void readLock() {
+            try {
+                readSemaphore.acquire();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException(ex);
+            }
+
+
+            if (readCount == 0) {
+                try {
+                    wrtieSemaphore.acquire();
+                } catch (InterruptedException e) {
+                    readSemaphore.release();
+                    throw new RuntimeException(e);
+                }
+            }
+
+            readCount++;
+            readSemaphore.release();
+        }
+
+        @Override
+        public void writeLock() {
+            try {
+                wrtieSemaphore.acquire();
```

Comment:
This method is called when data either needs to be loaded from the training index or when the allocation is closed. In the first case, it is called [here](https://github.com/jmazanec15/k-NN-1/blob/cache-refactor/src/main/java/org/opensearch/knn/index/memory/NativeMemoryLoadStrategy.java#L141). Because the allocation hasnt been returned to the cache yet, there is no case this would block.

In the second case, we need to acquire the write lock to close in order to prevent a free operation in JNI while the data is in use. When the data is needed for a read only operation (in this case, model training), the read lock would be obtained. 

The problem with this approach is that a thread would block while waiting to obtain the readLock during the loading of training data from the training index. I am not sure if there is a better way to handle this. The thread that actually will train the model will be in its own thread pool.

For index allocation, the write lock is only obtained when the allocation needs to be closed. Read locks are obtained during search. By default, I believe ReentrantLock uses non-fair mode, however, I could switch to fair mode if you think that would be better for index writes.

Reply:
> Because the allocation hasnt been returned to the cache yet, there is no case this would block.
Does it mean that since cache has no such entry yet, nobody will read it and thus no blocking?

> I am not sure if there is a better way to handle this. 
You can use tryAcquire.  If require fails, return immediately and schedule to tryAcquire again in some time (e.g., a few seconds if you expect that's the expected loading data time).  You can find examples in AD.

> By default, I believe ReentrantLock uses non-fair mode, however, I could switch to fair mode if you think that would be better for index writes.
I got no strong opinion on that.  The major difference is some delay (a few hundred milliseconds) if using fair mode.

Reply:
1. Yes. The write lock will be obtained before any other thread has a chance to get the read lock. Blocking may happen when a caller attempts to grab a read lock or if the allocation is closed (in which case the write lock needs to be obtained)

2. I see. So the benefit of this approach is that we could do something else with the thread while the permit cannot be acquired? I am not sure this applies for this case. For training, we are going to limit the number of training jobs per node to 1 (at least initially). Therefore, the training thread may block until the data is loaded (i.e. until the initial write lock is released), but this thread should not be performing any other training tasks at the same time anyway. That's why I think it is okay for this thread to block.

3. If we use acquire, I will set to fair mode.

Reply:
about 2.  What is the thread being blocked?  Is it a KNN thread used only for training? I found this read lock is in KNNWeight.scorer.  IDE points me a bunch of library calling this scorer method.  So I have no clue which thread is the caller.

Reply:
Right, so the readlock in KNNWeight is not for training, it is for searching indexes. In this case, it will only block when the write lock is obtained to close the index (thats the only time the write lock is obtained for the index allocation). Closing an index is a quick process, so this will not block very long. When it resumes, the search will fail.

The readlock for a training allocation will be called in a dedicated k-NN thread used for training a model.

Reply:
Could you point me the code link where a dedicated k-NN thread is used? According to https://opensourceconnections.com/blog/2014/01/20/build-your-own-custom-lucene-query-and-scorer/, it seems   KNNWeight.scorer will be invoked by some idxSearcher.search(q).

Reply:
Also, could you add your explanation to comments?

"Right, so the readlock in KNNWeight is not for training, it is for searching indexes. In this case, it will only block when the write lock is obtained to close the index (thats the only time the write lock is obtained for the index allocation). Closing an index is a quick process, so this will not block very long. When it resumes, the search will fail.

The readlock for a training allocation will be called in a dedicated k-NN thread used for training a model."

Reply:
Sure, dedicated k-NN thread is only relevant to training (which has not yet been implemented), so no code link.

The locks are on a per NativeMemoryAllocation basis (i.e. there is 1 lock per allocation). There are two NativeMemoryAllocation types: (1) IndexAllocation and (2) TrainingDataAllocation. At any given time, there may be several allocations in the NativeMemoryCacheManager's cache. Whenever you want to perform a read-only operation (such as search or training) on an allocation, you obtain the readlock. Whenever you want to perform an operation that modifies the allocation (such as deallocating the allocation or changing data in it), the write lock needs to be obtained.

IndexAllocation represents native library indices in native memory that can be searched. In order to search an index, the index first needs to be loaded into memory. The IndexAllocation represents this. On search, the readlock for this allocation needs to be grabbed. This happens in the KNNWeight function. The only case where this readlock will block is when the IndexAllocation is closed (meaning it is removed from native memory).

So, in [KNNWeight](https://github.com/jmazanec15/k-NN-1/blob/cache-refactor/src/main/java/org/opensearch/knn/index/KNNWeight.java#L129-L158), we first get an IndexAllocation from the cache. Then we grab the readlock on the allocation, then we search it, then we release the lock. To get the IndexAllocation from the NativeMemoryCacheManager, we create IndexEntryContext which has the functionality to get the Allocation from the cache if it is there, and, if it is not present, it contains the functionality to create the allocation in native memory and load the allocation into the cache. The only time the writelock is grabbed for an index allocation is [here](https://github.com/jmazanec15/k-NN-1/blob/cache-refactor/src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java#L277). So, the only case where KNNWeight will block is if the index is being deallocated.

Reply:
thanks for the explanation.  That's precise and clear.  Please add these comments in appropriate places.

I'd still suggest consider non-blocking way in training.  You can limit one training at one time using semaphore or locking.  You can free the thread to do sth else like training.  That will be scalable for future KNN extension.  Say tomorrow, you want to use the thread pool for other features.  Otherwise, you will need to add a threadpool for each new feature.

Reply:
For now, I think Im going to leave blocking. I think, for training, if we wanted to make it async, the logic here would require a small change (maybe an additional tryReadLock function). Most of the logic would be implemented in the training pool.